src/gui/AutomatableModelView.cpp

```diff
@@ -258,7 +279,236 @@ void AutomatableModelViewSlots::removeConnection()
+	if (clips->size() > 0)
+	{
+		// selecting the track with the most amount of clips
+		// connected to this model
+		AutomationTrack* maxTrack = dynamic_cast<AutomationTrack*>((*clips)[0]->getTrack());
+		int maxTrackCount = 1;
+		for (size_t i = 1; i < clips->size(); i++)
+		{
+			int currentCount = 0;
+			for (AutomationClip* j : (*clips))
+			{
+				if ((*clips)[i]->getTrack() == j->getTrack())
+				{
+					currentCount++;
+				}
+			}
+			if (maxTrackCount < currentCount)
+			{
+				maxTrackCount = currentCount;
+				maxTrack = dynamic_cast<AutomationTrack*>((*clips)[i]->getTrack());;
+			}
+		}
+		output = maxTrack;
+	}
+	else if (canAddNewTrack)
+	{
+		// adding new track
+		output = new AutomationTrack(getGUI()->songEditor()->m_editor->model(), false);
+	}
+	return output;
+}
+
+AutomationClip* AutomatableModelViewSlots::getCurrentAutomationClip(AutomationTrack* track, bool canAddNewClip, bool searchAfter)
+{
+	AutomationClip* output = nullptr;
+	const std::vector<Clip*>& trackClips = track->getClips();
+	TimePos timePos = static_cast<TimePos>(Engine::getSong()->getPlayPos());
+	
+	bool tryAdding = false;
+	if (trackClips.size() > 0)
+	{
+		// getting the closest clip that start before or after the global time position
+		tick_t closestTime = -1;
+		Clip* closestClip = nullptr;
+		for (Clip* currentClip : trackClips)
+		{
+			tick_t currentTime = currentClip->startPosition().getTicks();
+			bool smallerCheck = currentTime > closestTime || closestTime < 0;
+			bool biggerCheck = currentTime < closestTime || closestTime < 0;
+
+			if ((!searchAfter && smallerCheck && timePos.getTicks() > currentTime)
+				|| (searchAfter && biggerCheck && timePos.getTicks() <= currentTime))
+			{
+				closestTime = currentTime;
+				closestClip = currentClip;
+			}
+		}
+
+		// in some cases there could be no clips before or after the global time position
+		// if this is the case, try adding a new one
+		// (if this fails, return nullptr)
+		if (!closestClip)
+		{
+			tryAdding = true;
+		}
+		else
+		{
+			output = dynamic_cast<AutomationClip*>(closestClip);
+		}
+	}
+	else
+	{
+		tryAdding = true;
+	}
+	if (tryAdding && canAddNewClip)
+	{
+		// adding a new clip
+		output = makeNewClip(track, timePos, true);
+	}
+	return output;
+}
+
+const TimePos AutomatableModelViewSlots::getNearestAutomationNode(AutomationTrack* track, AutomationClip** clipOut)
+{
+	TimePos output;
+	AutomationClip* minClip = nullptr;
+	int minDistance = -1;
+
+	TimePos timePos = static_cast<TimePos>(Engine::getSong()->getPlayPos());
+	// getting the clips before and after the global time position
+	AutomationClip* clipBefore = getCurrentAutomationClip(track, false, false);
+	AutomationClip* clipAfter = getCurrentAutomationClip(track, false, true);
 
+	if (clipBefore && clipBefore->hasAutomation())
+	{
+		// getting nearest node
+		// in the clip that starts before this
+		for (auto it = clipBefore->getTimeMap().begin(); it != clipBefore->getTimeMap().end(); ++it)
```

Comment:
This loop can not be replaced, because `it` would be a pointer to a `typename T`node instead of a key.
Qt provides a `key()` function that finds a key for a given `T`, but that is not optimized.

Qt documentation:
> 
This function can be slow linear time, because QMap's internal data structure is optimized for fast lookup by key, not by value.

Reply:
Nvm, let this be like this.